lib/ddtrace/profiling/ext/cpu.rb

```diff
@@ -19,6 +20,65 @@ def self.apply!
           # will provide a thread/clock ID for CPU timing.
           require 'ddtrace/profiling/ext/cthread'
           ::Thread.send(:prepend, Profiling::Ext::CThread)
+
+          # Applying hooks will allow any existing threads to update
+          # their thread/clock IDs so their CPU timings can be measured.
+          apply_hooks!
+        end
+
+        def self.apply_hooks!
+          # Threads that have already been created, will not have resolved
+          # a thread/clock ID. This is because these IDs can only be resolved
+          # from within the thread's execution context, which we do not control.
+          #
+          # We can work around this by applying trace hooks to capture each thread
+          # from within its own execution context, resolving the thread/clock IDs,
+          # then disabling the hook so it doesn't re-run needlessly.
+          Thread.list.each do |thread|
+            # Only attempt to hook into threads that are instrumented
+            next unless thread.is_a?(CThread)
+
+            # TracePoint supports filtering by thread for Ruby 2.7+
+            if Gem::Version.new(RUBY_VERSION) >= Gem::Version.new('2.7') && defined?(TracePoint)
+              thread[THREAD_HOOK] = TracePoint.new(:line) do |tp|
```

Comment:
In the PR description you mention

> There is a number of events (e.g. :call, :line) that can be used to trigger the hook; some call more often than others. We probably want to be careful to call the right event(s) such that we call the hook quick enough, without being too aggressive.

I'm not entirely convinced this is the case for what we're doing. Because we know that we will want to run as soon as possible + only exactly once, why would hooking into every event be "aggressive" rather than "ASAP"?

Reply:
I'll have to think on this more; maybe it's okay.